### ISSUE TITLE
mbedtls_platform_entropy_poll() blocks

### ISSUE THREAD

**Issue by ashegoulding** [ARMmbed/mbedtls#3551]
### Description
- Type: Bug
- Priority: Minor

---------------------------------------------------------------
## Bug

**OS**  
Linux kernel version 3.17 or later

**mbed TLS build:**  
Version: 0ca6d38bc375cc8f1e6f99c896f98266eaef6a12

`mbedtls_platform_entropy_poll()` function calls `getrandom()` syscall if available and falls back to reading `/dev/urandom` if it's not. `getrandom()` call without `GRND_NONBLOCK` will block the calling thread until sufficient entropy is gathered on the system. This causes inconsistent behaviour: the function could block with kernel post-3.17 and will never block with pre-3.17.

References: random(7), getrandom(2)

The following patch is intended to change the function so that it will resort to reading `/dev/urandom` if `getrandom()` fails with **EAGAIN**. This is not safe, but applications shouldn't start so early after boot and the kernel warns about uninitialised `/dev/urandom` read, so...

```
diff --git a/library/entropy_poll.c b/library/entropy_poll.c
index 62fb4afbf..4d15e4411 100644
--- a/library/entropy_poll.c
+++ b/library/entropy_poll.c
@@ -96,6 +96,7 @@ int mbedtls_platform_entropy_poll( void *data, unsigned char *output, size_t len
 #include <sys/syscall.h>
 #if defined(SYS_getrandom)
 #define HAVE_GETRANDOM
+#include <sys/random.h>
 #include <errno.h>
 
 static int getrandom_wrapper( void *buf, size_t buflen, unsigned int flags )
@@ -157,15 +158,18 @@ int mbedtls_platform_entropy_poll( void *data,
     ((void) data);
 
 #if defined(HAVE_GETRANDOM)
-    ret = getrandom_wrapper( output, len, 0 );
+    ret = getrandom_wrapper( output, len, GRND_NONBLOCK );
     if( ret >= 0 )
     {
         *olen = ret;
         return( 0 );
     }
-    else if( errno != ENOSYS )
+    else if( errno != ENOSYS && errno != EAGAIN )
         return( MBEDTLS_ERR_ENTROPY_SOURCE_FAILED );
-    /* Fall through if the system call isn't known. */
+    /*
+     * Fall through if the system call isn't known
+     * or there isn't sufficient entropy.
+     */
 #else
     ((void) ret);
 #endif /* HAVE_GETRANDOM */
```

**Comment by gilles-peskine-arm** [comment]
> `getrandom()` call without `GRND_NONBLOCK` will block the calling thread until sufficient entropy is gathered on the system.

This is the desired behavior. `getrandom` without `GRND_NONBLOCK` is basically useless: it says “give me random bits, except if you can't do it quickly, give me non-random bits”. It emulates the broken behavior of `/dev/urandom` on Linux.

`getrandom(GRND_NONBLOCK)` is exactly what we want: get random bits, as fast as possible. If this involves blocking, it involves blocking.

> applications shouldn't start so early after boot

Mbed TLS gets used in all kind of applications, including ones that run early during the boot. And anyway, if it isn't the early boot, `getrandom(GRND_NONBLOCK)` won't block.

On Linux kernels before 3.16, without `getrandom`, there is unfortunately no sane way to obtain 
entropy. Reading from `/dev/urandom` succeeds without blocking but returns predictable data if the entropy pool hasn't been initialized. Reading from `/dev/random` is safe, but blocks so often that it isn't viable to use it systematically. I'm personally not happy about Mbed TLS using `/dev/urandom` in that case, but reading from `/dev/random` wouldn't be good either. So unfortunately, we're stuck between a rock and a hard place, and I don't have a good solution.

In summary, the inconsistency between pre- and post-3.16 is indeed a problem, but it's the pre-3.16 behavior that's wrong.

**Comment by ashegoulding** [comment]
Unacceptable. THIS IS A BUG. This is at least "undocumented".

All I need was to generate some random numbers, and, out of the blue, my entire program just hangs. Either you should document this behaviour or make the function behave the same way regardless of the kernel version. I agree that you need to make the library safe. But blocking the calling thread is just out of the question. Block or nonblock is a serious matter and I was not warned about this when I first read the manual.

Using `/dev/random` brings another problem. It imposes another security hole.

**On making it nonblocking**

I understand that this is difficult. You could check if `getrandom()` fails with `EAGAIN` and fail so the error can bubble all the way up to higher functions like `mbedtls_ctr_drbg_seed()`, but there's no way to tell if there's enough entropy when `getrandom()` is not available.

Most TLS implementations don't really care about the entropy anyways. When applications need random, you just have to give it to them because there's no way of knowing when the pool is going to be initialised. It could take minutes. Do you think blocking applications for minutes just because there's no entropy is acceptable??


**Workaround**

The workaround I chose was removing platform_entropy by hand by calling `mbedtls_entropy_init()`. However, the platform entropy(`/dev/urandom` or `getrandom()`) is the only strong source on most platforms. So I had to make a "fake" strong source so `mbedtls_entropy_gather()` wouldn't fail. I know this is not good, but it gets the job done.

```
static int urand_ent_f (void *data, unsigned char *output, size_t len, size_t *olen) {
	const int fd = open("/dev/urandom", O_RDONLY);
	int ret = 0;

	if (fd < 0 || read(fd, output, len) != (ssize_t)len) {
		ret = MBEDTLS_ERR_ENTROPY_SOURCE_FAILED;
	}
	*olen = len;

	if (fd >= 0) {
		close(fd);
	}

	return ret;
}

typedef struct {
	pid_t pid;
	pid_t ppid;
	clock_t clock;
	struct timespec now;
} ent_buf_t;

static int proc_ent_f (void *data, unsigned char *output, size_t len, size_t *olen) {
	ent_buf_t buf;

	memzero(&buf, sizeof(buf));
	buf.pid = getpid();
	buf.ppid = getppid();
	buf.clock = clock();
	clock_gettime(CLOCK_MONOTONIC, &buf.now);

	*olen = prne_op_min(len, sizeof(buf));
	memcpy(output, &buf, sizeof(*olen));

	return 0;
}

void my_mbedtls_entropy_init (mbedtls_entropy_context *ctx) {
	mbedtls_entropy_init(ctx);

	// Remove platform source, which could call getrandom()
	for (int i = 0; i < ctx->source_count; i += 1) {
		if (ctx->source[i].f_source == mbedtls_platform_entropy_poll) {
			memmove(ctx->source + i, ctx->source + i + 1, sizeof(mbedtls_entropy_source_state) * (ctx->source_count - i - 1));
			ctx->source_count -= 1;
			// Add our own implementation as the one just got removed could be the only source.
			mbedtls_entropy_add_source(ctx, urand_ent_f, NULL, MBEDTLS_ENTROPY_MIN_PLATFORM, MBEDTLS_ENTROPY_SOURCE_STRONG);
			mbedtls_entropy_add_source(ctx, proc_ent_f, NULL, sizeof(ent_buf_t), MBEDTLS_ENTROPY_SOURCE_STRONG);
			break;
		}
	}
}
```

**Comment by gilles-peskine-arm** [comment]
> Most TLS implementations don't really care about the entropy anyways.

Er, yes, we do. Without entropy on both sides of the connection, TLS is not secure. If only one party has entropy, the session key is safe (for RSA-without-DH ciphersuites, the server has to have entropy), but the session can be replayed.

> When applications need random, you just have to give it to them because there's no way of knowing when the pool is going to be initialised. It could take minutes. Do you think blocking applications for minutes just because there's no entropy is acceptable??

It's unavoidable. If you want random numbers, you need entropy. If you don't want random numbers, don't ask for them.

If it's a problem to block the thread that's calling the DRBG initialization, call it from a separate thread. Obviously you'll have to wait until the entropy is actually available to obtain random numbers.

**The code you posted is insecure**. It doesn't get the job done, it just pretends to do the job while leaving a security hole. If your application didn't block with an older kernel where Mbed TLS used `/dev/urandom`, and upgrading to a kernel with `getrandom` makes it block, you always had a security hole, you just didn't realize it.

If your application runs during the early boot and is blocked due to the lack of entropy, you have a platform integration problem. Potential solutions include:
* Redesign your boot process so that your application runs later.
* Early in the boot process, before anything needs entropy, load saved entropy from a file and update the saved entropy file. For the first boot, either accept the wait, or inject an entropy file from a trusted source during the manufacturing process.
* Add a hardware RNG to your platform. These days, even fairly low-end chips have a TRNG.

**Comment by ashegoulding** [comment]
@gilles-peskine-arm You're missing the point.

Just add a comment saying that the function could block. It just takes one line. And if you care so much about entropy, just make the function fail so users can fix their problem. Do you now how tedious it is to take a core dump from an embedded machine and run gdb on it from a PC?

**Comment by gilles-peskine-arm** [comment]
You mean add a note to the documentation of `mbedtls_entropy_func`, `mbedtls_entropy_gather` and `mbedtls_entropy_update_nv_seed` (I think these are the functions that can block, but I haven't checked very closely)? Sure, that would be a good idea. Feel free to submit a pull request for that.

**Comment by mpg** [comment]
I agree that we should document which functions can block.

I think we all agree that the patch you posted is not safe. Being a security library, we don't apply unsafe patches.

As Gilles said, the behaviour of `getrandom()` with no flags is exactly what we want, and there is a fairly strong consensus in the security community about that. This consensus is based on a number well-documented issues in the real world that occurred because not-so-random numbers were delivered when no actual entropy was available, see for example the [Mining your Ps and Qs](https://factorable.net/weakkeys12.extended.pdf) paper.

I understand you had a bad experience caused by the fact that this blocking was surprising to you, in part because it's new behaviour that wasn't present when running with older kernels, and in part because it's not documented.

Since there's no easy way to fix the behaviour on older kernels (which is one of the reasons `getrandom()` was introduced), I think the only thing we can do moving forward it indeed to document the potentially-blocking behaviour for each function that could block.

Regarding the larger picture, Gilles suggested 3 secure and practical ways to solve the issue, hopefully at least one on them will be applicable to your situation.

**Comment by mpg** [comment]
@ashegoulding Feel free to review the documentation changes I made in #3573 and let us know if they address your concerns with the documentation.

**Comment by ashegoulding** [comment]
Top job
